File: typescript/src/api/types/order.ts
/**
 * Order-related types for the Lightcone REST API.
 */

/**
 * Order side enum.
 */
export type ApiOrderSide = 0 | 1;

/**
 * Order status values.
 */
export type OrderStatusValue = "accepted" | "partial_fill" | "filled" | "rejected";

/**
 * Fill information from order matching.
 */
export interface Fill {
  /** Counterparty address */
  counterparty: string;
  /** Counterparty's order hash */
  counterparty_order_hash: string;
  /** Amount filled as decimal string */
  fill_amount: string;
  /** Fill price as decimal string */
  price: string;
  /** Whether this order was the maker */
  is_maker: boolean;
}

/**
 * Request for POST /api/orders/submit.
 */
export interface SubmitOrderRequest {
  /** Order creator's pubkey (Base58) */
  maker: string;
  /** User's nonce for uniqueness */
  nonce: number;
  /** Market address (Base58) */
  market_pubkey: string;
  /** Token being bought/sold (Base58) */
  base_token: string;
  /** Token used for payment (Base58) */
  quote_token: string;
  /** Order side (0=BID, 1=ASK) */
  side: number;
  /** Amount maker gives */
  maker_amount: number;
  /** Amount maker wants to receive */
  taker_amount: number;
  /** Unix timestamp, 0=no expiration */
  expiration?: number;
  /** Ed25519 signature (hex, 128 chars) */
  signature: string;
  /** Target orderbook */
  orderbook_id: string;
}

/**
 * Response for POST /api/orders/submit.
 */
export interface OrderResponse {
  /** Order hash (hex) */
  order_hash: string;
  /** Order status */
  status: string;
  /** Remaining amount as decimal string */
  remaining: string;
  /** Filled amount as decimal string */
  filled: string;
  /** Fill details */
  fills: Fill[];
}

/**
 * Request for POST /api/orders/cancel.
 */
export interface CancelOrderRequest {
  /** Hash of order to cancel (hex) */
  order_hash: string;
  /** Must match order creator (Base58) */
  maker: string;
}

/**
 * Response for POST /api/orders/cancel.
 */
export interface CancelResponse {
  /** Cancellation status */
  status: string;
  /** Order hash */
  order_hash: string;
  /** Remaining amount that was cancelled as decimal string */
  remaining: string;
}

/**
 * Request for POST /api/orders/cancel-all.
 */
export interface CancelAllOrdersRequest {
  /** User's public key (Base58) */
  user_pubkey: string;
  /** Limit to specific market (empty = all) */
  market_pubkey?: string;
}

/**
 * Response for POST /api/orders/cancel-all.
 */
export interface CancelAllResponse {
  /** Status (success) */
  status: string;
  /** User pubkey */
  user_pubkey: string;
  /** Market pubkey if specified */
  market_pubkey?: string;
  /** List of cancelled order hashes */
  cancelled_order_hashes: string[];
  /** Count of cancelled orders */
  count: number;
  /** Human-readable message */
  message: string;
}

/**
 * User order from GET /api/users/orders.
 */
export interface UserOrder {
  /** Order hash */
  order_hash: string;
  /** Market pubkey */
  market_pubkey: string;
  /** Orderbook ID */
  orderbook_id: string;
  /** Order side (0=BID, 1=ASK) */
  side: number;
  /** Maker amount as decimal string */
  maker_amount: string;
  /** Taker amount as decimal string */
  taker_amount: string;
  /** Remaining amount as decimal string */
  remaining: string;
  /** Filled amount as decimal string */
  filled: string;
  /** Order price as decimal string */
  price: string;
  /** Creation timestamp */
  created_at: string;
  /** Expiration timestamp */
  expiration: number;
}

/**
 * Request for POST /api/users/orders.
 */
export interface GetUserOrdersRequest {
  /** User's public key (Base58) */
  user_pubkey: string;
}

/**
 * Outcome balance in user orders response.
 */
export interface UserOrderOutcomeBalance {
  /** Outcome index */
  outcome_index: number;
  /** Conditional token address */
  conditional_token: string;
  /** Idle balance as decimal string */
  idle: string;
  /** Balance on order book as decimal string */
  on_book: string;
}

/**
 * User balance from GET /api/users/orders.
 */
export interface UserBalance {
  /** Market pubkey */
  market_pubkey: string;
  /** Deposit asset */
  deposit_asset: string;
  /** Outcome balances */
  outcomes: UserOrderOutcomeBalance[];
}

/**
 * Response for POST /api/users/orders.
 */
export interface UserOrdersResponse {
  /** User pubkey */
  user_pubkey: string;
  /** Open orders */
  orders: UserOrder[];
  /** User balances */
  balances: UserBalance[];
}

File: rust/src/api/types/order.rs
//! Order-related types for the Lightcone REST API.

use serde::{Deserialize, Serialize};

/// Order side enum (serializes as integer: 0=Bid, 1=Ask).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(try_from = "u32", into = "u32")]
#[repr(u32)]
pub enum ApiOrderSide {
    /// Buy base token with quote token
    Bid = 0,
    /// Sell base token for quote token
    Ask = 1,
}

/// Error returned when trying to convert an invalid value to ApiOrderSide.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct InvalidOrderSideError(pub u32);

impl std::fmt::Display for InvalidOrderSideError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "invalid order side value: {} (expected 0 for Bid or 1 for Ask)", self.0)
    }
}

impl std::error::Error for InvalidOrderSideError {}

impl TryFrom<u32> for ApiOrderSide {
    type Error = InvalidOrderSideError;

    fn try_from(value: u32) -> Result<Self, Self::Error> {
        match value {
            0 => Ok(Self::Bid),
            1 => Ok(Self::Ask),
            _ => Err(InvalidOrderSideError(value)),
        }
    }
}

impl From<ApiOrderSide> for u32 {
    fn from(side: ApiOrderSide) -> Self {
        side as u32
    }
}

/// Order status enum.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum OrderStatus {
    /// Order placed on book
    Accepted,
    /// Partially filled, remainder on book
    PartialFill,
    /// Completely filled
    Filled,
    /// Order rejected
    Rejected,
}

/// Fill information from order matching.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Fill {
    /// Counterparty address
    pub counterparty: String,
    /// Counterparty's order hash
    pub counterparty_order_hash: String,
    /// Amount filled as decimal string
    pub fill_amount: String,
    /// Fill price as decimal string
    pub price: String,
    /// Whether this order was the maker
    pub is_maker: bool,
}

/// Request for POST /api/orders/submit.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SubmitOrderRequest {
    /// Order creator's pubkey (Base58)
    pub maker: String,
    /// User's nonce for uniqueness
    pub nonce: u64,
    /// Market address (Base58)
    pub market_pubkey: String,
    /// Token being bought/sold (Base58)
    pub base_token: String,
    /// Token used for payment (Base58)
    pub quote_token: String,
    /// Order side (0=BID, 1=ASK)
    pub side: u32,
    /// Amount maker gives
    pub maker_amount: u64,
    /// Amount maker wants to receive
    pub taker_amount: u64,
    /// Unix timestamp, 0=no expiration
    #[serde(default)]
    pub expiration: i64,
    /// Ed25519 signature (hex, 128 chars)
    pub signature: String,
    /// Target orderbook
    pub orderbook_id: String,
}

/// Response for POST /api/orders/submit.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrderResponse {
    /// Order hash (hex)
    pub order_hash: String,
    /// Order status
    pub status: OrderStatus,
    /// Remaining amount as decimal string
    pub remaining: String,
    /// Filled amount as decimal string
    pub filled: String,
    /// Fill details
    #[serde(default)]
    pub fills: Vec<Fill>,
}

/// Request for POST /api/orders/cancel.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CancelOrderRequest {
    /// Hash of order to cancel (hex)
    pub order_hash: String,
    /// Must match order creator (Base58)
    pub maker: String,
}

/// Response for POST /api/orders/cancel.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CancelResponse {
    /// Cancellation status
    pub status: String,
    /// Order hash
    pub order_hash: String,
    /// Remaining amount that was cancelled as decimal string
    pub remaining: String,
}

/// Request for POST /api/orders/cancel-all.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CancelAllOrdersRequest {
    /// User's public key (Base58)
    pub user_pubkey: String,
    /// Limit to specific market (empty = all)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub market_pubkey: Option<String>,
}

/// Response for POST /api/orders/cancel-all.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CancelAllResponse {
    /// Status (success)
    pub status: String,
    /// User pubkey
    pub user_pubkey: String,
    /// Market pubkey if specified
    #[serde(default)]
    pub market_pubkey: Option<String>,
    /// List of cancelled order hashes
    pub cancelled_order_hashes: Vec<String>,
    /// Count of cancelled orders
    pub count: u64,
    /// Human-readable message
    pub message: String,
}

/// User order from GET /api/users/orders.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserOrder {
    /// Order hash
    pub order_hash: String,
    /// Market pubkey
    pub market_pubkey: String,
    /// Orderbook ID
    pub orderbook_id: String,
    /// Order side
    pub side: ApiOrderSide,
    /// Maker amount as decimal string
    pub maker_amount: String,
    /// Taker amount as decimal string
    pub taker_amount: String,
    /// Remaining amount as decimal string
    pub remaining: String,
    /// Filled amount as decimal string
    pub filled: String,
    /// Order price as decimal string
    pub price: String,
    /// Creation timestamp
    pub created_at: String,
    /// Expiration timestamp
    pub expiration: i64,
}

/// Request for POST /api/users/orders.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GetUserOrdersRequest {
    /// User's public key (Base58)
    pub user_pubkey: String,
}

/// Outcome balance in user orders response.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserOrderOutcomeBalance {
    /// Outcome index
    pub outcome_index: u32,
    /// Conditional token address
    pub conditional_token: String,
    /// Idle balance as decimal string
    pub idle: String,
    /// Balance on order book as decimal string
    pub on_book: String,
}

/// User balance from GET /api/users/orders.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserBalance {
    /// Market pubkey
    pub market_pubkey: String,
    /// Deposit asset
    pub deposit_asset: String,
    /// Outcome balances
    pub outcomes: Vec<UserOrderOutcomeBalance>,
}

/// Response for POST /api/users/orders.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserOrdersResponse {
    /// User pubkey
    pub user_pubkey: String,
    /// Open orders
    pub orders: Vec<UserOrder>,
    /// User balances
    pub balances: Vec<UserBalance>,
}
